File: roberto/solutions/correct/nlehmann.cpp
#include <algorithm>
#include <climits>
#include <cstdio>
#include <set>
#include <tuple>
#include <vector>
using namespace std;

pair<vector<int>, vector<int>> compute_min_max_arrs(const vector<int> &arr,
                                                    int w) {
  vector<int> mins(arr.size() - w + 1);
  vector<int> maxs(arr.size() - w + 1);
  multiset<int> elements;

  for (int i = 0; i < w; ++i) {
    elements.insert(arr[i]);
  }
  mins[0] = *elements.begin();
  maxs[0] = *elements.rbegin();

  for (int i = 1; i + w - 1 < arr.size(); ++i) {
    elements.erase(elements.find(arr[i - 1]));
    elements.insert(arr[i + w - 1]);
    mins[i] = *elements.begin();
    maxs[i] = *elements.rbegin();
  }

  return {mins, maxs};
}

int main() {
  int H, W, M, N;
  scanf("%d%d%d%d", &H, &W, &M, &N);

  vector<vector<int>> terrain(H, vector<int>(W));

  for (int i = 0; i < H; ++i) {
    for (int j = 0; j < W; ++j) {
      scanf("%d", &terrain[i][j]);
    }
  }

  vector<vector<int>> mins_for_row(H);
  vector<vector<int>> maxs_for_row(H);
  for (int i = 0; i < H; ++i) {
    tie(mins_for_row[i], maxs_for_row[i]) = compute_min_max_arrs(terrain[i], N);
  }

  int best = INT_MAX;
  for (int j = 0; j + N - 1 < W; ++j) {
    multiset<int> elements_min;
    multiset<int> elements_max;

    for (int i = 0; i < M; ++i) {
      elements_min.insert(mins_for_row[i][j]);
      elements_max.insert(maxs_for_row[i][j]);
    }
    best = min(best, *elements_max.rbegin() - *elements_min.begin());
    for (int i = 1; i + M - 1 < H; ++i) {
      elements_min.erase(elements_min.find(mins_for_row[i - 1][j]));
      elements_min.insert(mins_for_row[i + M - 1][j]);
      elements_max.erase(elements_max.find(maxs_for_row[i - 1][j]));
      elements_max.insert(maxs_for_row[i + M - 1][j]);

      best = min(best, *elements_max.rbegin() - *elements_min.begin());
    }
  }
  printf("%d\n", best);
}

File: roberto/solutions/correct/nlehmann.java
import java.util.*;
import java.io.*;

class roberto {
  public static void main(String[] args) throws Exception {
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(in.readLine());

    int H = Integer.parseInt(st.nextToken());
    int W = Integer.parseInt(st.nextToken());
    int M = Integer.parseInt(st.nextToken());
    int N = Integer.parseInt(st.nextToken());

    int terrain[][] = new int[H][W];
    for (int i = 0; i < H; ++i) {
      st = new StringTokenizer(in.readLine());
      for (int j = 0; j < W; ++j) {
        terrain[i][j] = Integer.parseInt(st.nextToken());
      }
    }

    int mins_for_row[][] = new int[H][];
    int maxs_for_row[][] = new int[H][];
    for (int i = 0; i < H; ++i) {
      Pair<int[]> p = compute_min_max_arrs(terrain[i], N);
      mins_for_row[i] = p.x;
      maxs_for_row[i] = p.y;
    }

    int best = Integer.MAX_VALUE;
    for (int j = 0; j + N - 1 < W; ++j) {
      TreeMap<Integer, Integer> elements_min = new TreeMap<>();
      TreeMap<Integer, Integer> elements_max = new TreeMap<>();

      for (int i = 0; i < M; ++i) {
        insert(elements_min, mins_for_row[i][j]);
        insert(elements_max, maxs_for_row[i][j]);
      }
      best = Math.min(best, elements_max.lastKey() - elements_min.firstKey());
      for (int i = 1; i + M - 1 < H; ++i) {
        remove(elements_min, mins_for_row[i - 1][j]);
        insert(elements_min, mins_for_row[i + M - 1][j]);

        remove(elements_max, maxs_for_row[i - 1][j]);
        insert(elements_max, maxs_for_row[i + M - 1][j]);

        best = Math.min(best, elements_max.lastKey() - elements_min.firstKey());
      }
    }

    System.out.println(best);
  }

  static Pair<int[]> compute_min_max_arrs(int arr[], int w) {
    int mins[] = new int[arr.length - w + 1];
    int maxs[] = new int[arr.length - w + 1];
    TreeMap<Integer, Integer> elements = new TreeMap<>();
    for (int i = 0; i < w; ++i) {
      insert(elements, arr[i]);
    }
    mins[0] = elements.firstKey();
    maxs[0] = elements.lastKey();

    for (int i = 1; i + w - 1 < arr.length; ++i) {
      remove(elements, arr[i - 1]);
      insert(elements, arr[i + w - 1]);
      mins[i] = elements.firstKey();
      maxs[i] = elements.lastKey();
    }
    return new Pair(mins, maxs);
  }

  static void insert(TreeMap<Integer, Integer> set, int element) {
    int n = set.containsKey(element) ? set.get(element) : 0;
    set.put(element, n + 1);
  }

  static void remove(TreeMap<Integer, Integer> set, int element) {
    if (set.containsKey(element)) {
      int n = set.get(element);
      if (n == 1) {
        set.remove(element);
      } else {
        set.put(element, n - 1);
      }
    }
  }
}

class Pair<T> {
  T x, y;

  public Pair(T x, T y) {
    this.x = x;
    this.y = y;
  }
}
